### PR TITLE
Support include/exclude regexes for event and resource constraints

### LINKED ISSUE #954: Muting certain events
## Overview

I am looking for a way to mute certain events at the source level. I found that I can write a regex for `event.message` to filter the events, but the Go regex engine does not allow for negative lookahead/lookbehind, so writing and maintaining the right regex to allow everything except what I want to filter becomes a mess.

It would be nice if I could write a negative filter, to allow anything except for what is matched by the criteria.

## Acceptance Criteria

- Allow negative filters on `event.message` and `event.kind`.
- When the criteria matches, the event is ignored, when the criteria does not match the event is processed.

## Reason

We need a way to ignore certain cluster events, but allow everything else.

### PR DESCRIPTION
**Pull request by pkosiec**
## Description

Changes proposed in this pull request:

- Support `include/exclude` regexes for `event.reason`, `event.message` and `resources[].name`.

## Breaking changes

_(To copy and paste in https://github.com/kubeshop/botkube/issues/972)_

- Kubernetes source constraints `event.reason`, `event.message` and `resources[].name` are now objects with two lists: `include` and `exclude`. Both take exact values or regex patterns. Exclude takes precedence over include.

   This allows you to set `exclude` patterns, and also unifies the configuration with `namespaces` constraints.

## Notes - open question
I didn't rename `namespaces` to `namespace` or `event.types` to `event.type` as I weren't sure if that's a good change. However, if you have any ideas, then please let me know - we're already doing a breaking change here, so I can unify the config properties more 🙂 

Maybe we should rename `namespaces` to `namespace` for consistency (as it is an object), and keep plural form for `event.types` as it is an array? Let me know 🙂 

## TODO

- [x] Update docs: https://github.com/kubeshop/botkube-docs/pull/211

## Testing

Check out the PR (`gh pr checkout 963`).

Use the configuration:
```
communications:
  "default-group":
    socketSlack:
      enabled: true
      appToken: "xapp-"
      botToken: "xoxb-"
      notification:
        type: "short"
      channels:
        "default":
          name: botkube-demo
          bindings:
            executors:
              - kubectl-read-only
              - helm
              - plugin-based
            sources:
              - k8s-all-events
              - k8s-recommendation-events
              - plugin-based
        "secondary":
          name: priv-channel
          bindings:
            executors: []
            sources:
              - k8s-all-events-except
sources:
  'k8s-all-events-except':
    displayName: "Kubernetes Info"
    kubernetes:
      namespaces:
        include:
          - ".*"
      event:
        types:
          - create
          - delete
          - error
        message:
          include: [".*"]
          exclude: ["^Readiness probe failed.*$"]
      resources:
        - type: v1/pods
        - type: v1/services
        - type: networking.k8s.io/v1/ingresses
        - type: v1/nodes
        - type: v1/namespaces
        - type: v1/persistentvolumes
        - type: v1/persistentvolumeclaims
        - type: v1/configmaps
        - type: rbac.authorization.k8s.io/v1/roles
        - type: rbac.authorization.k8s.io/v1/rolebindings
        - type: rbac.authorization.k8s.io/v1/clusterrolebindings
        - type: rbac.authorization.k8s.io/v1/clusterroles
        - type: apps/v1/daemonsets
          event:
            types:
              - create
              - update
              - delete
              - error
          updateSetting:
            includeDiff: true
            fields:
              - spec.template.spec.containers[*].image
              - status.numberReady
        - type: batch/v1/jobs
          event:
            types:
              - create
              - update
              - delete
              - error
          updateSetting:
            includeDiff: true
            fields:
              - spec.template.spec.containers[*].image
              - status.conditions[*].type
        - type: apps/v1/deployments
          event:
            types:
              - create
              - update
              - delete
              - error
          updateSetting:
            includeDiff: true
            fields:
              - spec.template.spec.containers[*].image
              - status.availableReplicas
        - type: apps/v1/statefulsets
          event:
            types:
              - create
              - update
              - delete
              - error
          updateSetting:
            includeDiff: true
            fields:
              - spec.template.spec.containers[*].image
              - status.readyReplicas

settings:
  clusterName: "dev"
  upgradeNotifier: false
```

Run Botkube:
```
export BOTKUBE_SETTINGS_LOG_LEVEL=info
export BOTKUBE_PLUGINS_CACHE__DIR="/tmp/plugins"
export BOTKUBE_SETTINGS_KUBECONFIG=$KUBECONFIG # optional
export BOTKUBE_CONFIG_PATHS="$(pwd)/helm/botkube/values.yaml,$(pwd)/comm_config.yaml"
go run cmd/botkube/main.go
```

Run the command to trigger events:
```
k delete po -n kube-system -l k8s-app=metrics-server
```

See events with `Readiness probe failed` message:
image

On `priv-channel`, they will be filtered out nicely:

image

## Related issue(s)

#954

### CODE REVIEW

**Comment by pkosiec** [comment]
Converting to draft after internal conversation about our preferred approach. The agreement is to rework the PR to:
- remove the advanced regex engine,
- introduce `include`/`exclude` lists in all of the source filters and use Go regexes there.

Converting this PR to draft then 🙂

File: internal/source/registration.go [hunk 1]
```diff
@@ -145,81 +144,104 @@ func (r registration) sourcesForEvent(routes []route, event event.Event) ([]stri
...
 		}
 
-		// labels
-		if !kvsSatisfiedForMap(route.labels, event.ObjectMeta.Labels) {
-			continue
+		for _, msg := range eventMsgs {
+			match, err := route.event.Message.IsAllowed(msg)
+			if err != nil {
+				lastErr = err
+				continue
```
Comment (josefkarasek) [inline internal/source/registration.go:192]: What's the purpose of `lastErr`? Why wait for the last error and not break on first error?

**Review by josefkarasek — APPROVED** [review]
lgtm :+1:

File: internal/source/registration.go (hunk 1 above)
Comment (pkosiec) [reply internal/source/registration.go:192]: 🤔 Right... It will be always the same error about compiling the same regex. Good catch, will fix that!

**Pull request merged**